### ISSUE TITLE
Haproxy 2.8.5: The health check to backend server is not sent using the VIP (Virtual IP) as the source IP address

### ISSUE THREAD

**Issue by SM-9870** [haproxy/haproxy#2781]
### Detailed Description of the Problem

After we migrate from haproxy 2.0 to 2.8.5, we are seeing a change in behavior for http checks.

The health checks to backend servers are not using the VIP as the source IP address whereas this was working fine in version 2.0

We are using config as below:

```
backend default
  mode tcp
  balance roundrobin
  source 192.x.x.x
  option httpchk GET /Test HTTP/1.2
  http-check send hdr Host <dns>
  http-check expect status 200
  default-server check port 443 send-proxy-v2 check-ssl verify required ca-file /etc/ssl/certs/ca-certificates.crt check-send-proxy resolvers defaultdns check-sni <dns> inter 60s fall 3 rise 2 fastinter 1s resolve-prefer ipv4

  server 10.x.x.x 10.x.x.x 
  server 10.x.x.x 10.x.x.x 
  server 10.x.x.x 10.x.x.x 
```

### Expected Behavior

The http checks should pass the client source IP as VIP. The backend server should see the http check traffic with source as VIP.

### Steps to Reproduce the Behavior

Use the default haproxy config as below:

```
backend default
  mode tcp
  balance roundrobin
  source 192.x.x.x
  option httpchk GET /Test HTTP/1.2
  http-check send hdr Host <dns>
  http-check expect status 200
  default-server check port 443 send-proxy-v2 check-ssl verify required ca-file /etc/ssl/certs/ca-certificates.crt check-send-proxy resolvers defaultdns check-sni <dns> inter 60s fall 3 rise 2 fastinter 1s resolve-prefer ipv4

  server 10.x.x.x 10.x.x.x 
  server 10.x.x.x 10.x.x.x 
  server 10.x.x.x 10.x.x.x 
```

### What is your configuration?

```haproxy
backend default
  mode tcp
  balance roundrobin
  source 192.x.x.x
  option httpchk GET /Test HTTP/1.2
  http-check send hdr Host <dns>
  http-check expect status 200
  default-server check port 443 send-proxy-v2 check-ssl verify required ca-file /etc/ssl/certs/ca-certificates.crt check-send-proxy resolvers defaultdns check-sni <dns> inter 60s fall 3 rise 2 fastinter 1s resolve-prefer ipv4

  server 10.x.x.x 10.x.x.x 
  server 10.x.x.x 10.x.x.x 
  server 10.x.x.x 10.x.x.x 
```
```


### Output of `haproxy -vv`

```plain
HAProxy version 2.8.5-1ubuntu3 2024/04/01 - https://haproxy.org/
Status: long-term supported branch - will stop receiving fixes around Q2 2028.
Known bugs: http://www.haproxy.org/bugs/bugs-2.8.5.html
```

**Comment by Darlelet** [comment]
Hi

I'm unable to reproduce with the infos provided. Do you mean haproxy ignores the `source` keyword when performing healthchecks?

Also, please consider updating to the latest 2.8 because 2.8.5 is affected by 264 known bugs: http://www.haproxy.org/bugs/bugs-2.8.5.html

**Comment by Darlelet** [comment]
Also if you want to present input client IP to backend servers, you need to use `source + usesrc` combination

**Comment by SM-9870** [comment]
We upgraded to latest version 3.x but still seeing the same behavior. 
Also even with source + usesrc, its not working.

Its the only problem with the http-checks otherwise for other traffic, we are able to see the client IPs in the backend side. 

The real problem is - we are expecting the VIP (192.x.x.) as client IP in the http-check (Layer 7 GET API). 

Also we noticed that when we use `send-proxy` its working as expected but the problem is with `send-proxy-v2`. 


Can you please confirm if there is any behavior change starting version 2.2 ? I tried to look into the changelog but did not find any info.

**Comment by Darlelet** [comment]
>The real problem is - we are expecting the VIP (192.x.x.) as client IP in the http-check (Layer 7 GET API).

Sorry but I don't understand, how are you checking the IP in your application? Are you looking at a specific header in the http payload or are you relying on actual IP address at layer 4?

`source` will only help to set the source IP address (layer 4) as seen on the wire when inspecting tcp packets

>Can you please confirm if there is any behavior change starting version 2.2 ? I tried to look into the changelog but did not find any info.

I couldn't find anything either, but it would really help to know more about your setup in order to check if it's expected or not. I tested the `source` keyword on 2.2 and 2.8 and couldn't find a change of behavior

**Comment by Darlelet** [comment]
Ok, now focusing on `send-proxy`, since it seems to be what you are interested about for transporting the client IP:

There have been some refacto arounds health checks in 2.2, so perhaps something changed between 2.0 and 2.2 indeed, I'll test

**Comment by Darlelet** [comment]
There is something I still don't understand though:

>The http checks should pass the client source IP as VIP. The backend server should see the http check traffic with source as VIP.

The healthchecks are performed alongside the traffic by haproxy itself. Unless you use `source + usesrc` to force a specific address, it will be haproxy's address. I don't see how you could expect the source address (whether it's checked at layer 4 or later using PROXY protocol) to match the one from the clients since the health check requests don't originate from the client (unlike regular traffic)

**Comment by capflam** [comment]
If it is an issue with the PROXY protocol V2, and you observe a difference between 2.0 and newer versions, it may be because there is a bug in older versions (2.1 and lower). Local connections, initiated by HAProxy, are not handled the same way than "proxified"  connections. But it is not properly handled on 2.1 and lower. It was fixed on the 2.2.

Take a look to the `pp2-never-send-local` global option.

**Comment by capflam** [comment]
I'm closing because it seems the issue can be fixed via a configuration directive. If I'm wrong, please reopen the issue. Thanks !
